day8/alienalphabet.py

```diff
@@ -0,0 +1,131 @@
+##############
+## Solution ##
+##############
+
+def alienAlphabet(orderedWords):
+    # edge cases
+    if len(orderedWords) <= 0:
+        return ''
+    if len(orderedWords) == 1:
+        return removeDups(orderedWords[0])
+
+    # linearization of a DAG
+    graph = createDependencyGraph(orderedWords)
+    alphabet = topSort(graph)
+
+    return "".join(alphabet)
+
+#############
+## Helpers ##
+#############
+
+def createDependencyGraph(orderedWords):
+    graph = {}
+    for i, word in enumerate(orderedWords[:-1]):
+        nextWord = orderedWords[i+1]
+        shorter = min(len(word), len(nextWord))
+        for j in xrange(shorter):
+            letter = word[j]
+            otherLetter = nextWord[j]
+
+            if letter not in graph:
+                graph[letter] = []
+            if otherLetter not in graph:
+                graph[otherLetter] = []
+
+            if letter != otherLetter:
+                graph[letter].append(otherLetter)
+                break
+
+    return graph
+
+# graph is represented as key:value pair
+# key = node
+# value = [list, of, connected, nodes]
+#
+# the graph is also given to be a DAG so all the connected
+# nodes are from outgoing edges
+#
+# this function returns the resulting linearization in a list
+def topSort(graph):
+    sources = getSources(graph)
+    visited = set()
+    postOrderStack = []
+
+    def dfs(start, graph, visitedSet, postOrderStack):
+        visitedSet.add(start)
+
+        # recurse
+        for child in graph[start]:
+            if child not in visitedSet:
+                dfs(child, graph, visitedSet, postOrderStack)
+
+        # done exploring your children, add yourself to the postorder
+        postOrderStack.append(start)        
+
+    # run dfs from all source nodes
+    for node in sources:
+        dfs(node, graph, visited, postOrderStack)
+
+    return postOrderStack[::-1]
+
+def getSources(graph):
```

Comment:
So as it turns out you actually don't need to figure out the sources. You can just start dfs on any node, and once you've explored all the things that depend on it, you can safely add it to the list since everything that depends on it has already been added (I cheated and read this from wikipedia)